src/librustdoc/html/highlight.rs

```diff
@@ -437,13 +523,68 @@ fn exit_span(out: &mut Buffer) {
 ///     enter_span(Foo), string("text", None), exit_span()
 ///     string("text", Foo)
 /// ```
+///
 /// The latter can be thought of as a shorthand for the former, which is more
 /// flexible.
-fn string<T: Display>(out: &mut Buffer, text: T, klass: Option<Class>) {
-    match klass {
-        None => write!(out, "{}", text),
-        Some(klass) => write!(out, "<span class=\"{}\">{}</span>", klass.as_html(), text),
+///
+/// Note that if `context` is not `None` and that the given `klass` contains a `Span`, the function
+/// will then try to find this `span` in the `span_correspondance_map`. If found, it'll then
+/// generate a link for this element (which corresponds to where its definition is located).
+fn string<T: Display>(
+    out: &mut Buffer,
+    text: T,
+    klass: Option<Class>,
+    context_info: &Option<ContextInfo<'_, '_, '_>>,
+) {
+    let klass = match klass {
+        None => return write!(out, "{}", text),
+        Some(klass) => klass,
+    };
+    if let Some(def_span) = klass.get_span() {
+        let mut text = text.to_string();
+        if text.contains("::") {
+            text = text.split("::").intersperse("::").fold(String::new(), |mut path, t| {
+                match t {
+                    "self" | "Self" => write!(
+                        &mut path,
+                        "<span class=\"{}\">{}</span>",
+                        Class::Self_(LightSpan::empty()).as_html(),
+                        t
+                    ),
+                    "crate" | "super" => write!(
+                        &mut path,
+                        "<span class=\"{}\">{}</span>",
+                        Class::KeyWord.as_html(),
+                        t
+                    ),
+                    t => write!(&mut path, "{}", t),
+                }
+                .expect("Failed to build source HTML path");
+                path
+            });
+        }
+        if let Some(context_info) = context_info {
+            if let Some(href) =
+                context_info.context.shared.span_correspondance_map.get(&def_span).and_then(
+                    |href| {
+                        let context = context_info.context;
+                        match href {
+                            LinkFromSrc::Local(span) => context
+                                .href_from_span(*span)
+                                .map(|s| format!("{}{}", context_info.root_path, s)),
+                            LinkFromSrc::External(def_id) => {
+                                format::href(*def_id, context).map(|(url, _, _)| url)
```

Comment:
Because we don't have access to their source map. Another point: the external crate might not have generated its source code. So instead, we simply link to the documentation page where you can access the source code if you want to (and if available). A next step that is planned is to have both links to an item documentation and source code in small "popup" when hovering the link or something along the line.

Reply:
> Because we don't have access to their source map.

We shouldn't need it. `href_from_span` can generate links to source files of external crates with a span from `tcx.def_span(def_id)`. That's how the [src] link on https://doc.rust-lang.org/nightly/std/clone/trait.Clone.html works for example.

> the external crate might not have generated its source code.

That's #55957. The solution is to check if the external crate has `#![doc(html_no_source)]`.

> A next step that is planned is to have both links to an item documentation and source code in small "popup" when hovering the link or something along the line.

That sounds like a good idea.

Reply:
> That's #55957. The solution is to check if the external crate has #![doc(html_no_source)].

Well, that's in case the crate was generated and we are not using external URL. :3

Unless I'm missing something, we never generated `[src]` links to extern items' definition, only links to extern items' documentation.

Reply:
> Unless I'm missing something, we never generated `[src]` links to extern items' definition

We absolutely do. Currently the `[src]` like on https://doc.rust-lang.org/nightly/std/clone/trait.Clone.html points to https://doc.rust-lang.org/nightly/src/core/clone.rs.html#108-133. Notice how that's a page in the `std` docs pointing directly to a source file in `core`. The code that handles linking to source pages of external crates is [here](https://github.com/rust-lang/rust/blob/60f1a2fc4b535ead9c85ce085fdce49b1b097531/src/librustdoc/html/render/context.rs#L315-L338). That code will work just fine here.

However the more important issue is that for links to external items this is still not taking into account `root_path` which means this is currently generating broken links. For example when enabling this on the `std` docs `doc/src/alloc/alloc.rs.html` contains links like `<a class="ident" href="core/ptr/non_null/struct.NonNull.html">NonNull</a>` which should be `<a class="ident" href="../../core/ptr/non_null/struct.NonNull.html">NonNull</a>`.

Reply:
Ah right, I forgot about the reexports... Thanks for bringing it up!

Reply:
This is now fixed, I also added an external type to the test to ensure it was correctly generated. Great catch, thanks again!